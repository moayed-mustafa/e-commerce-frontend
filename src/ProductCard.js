

import React, { useState, useContext} from 'react'
import './index.css'
import ServerApi from './API/server'
import { useSelector, useDispatch } from 'react-redux'
import { Alert } from 'reactstrap'
import userContext from './userContext'
import { v4 as uuid } from 'uuid';


export default function ProductCard({ product }) {

    const [flash, setFlash] = useState({condition:false, message:"", backgroundColor: ""})

    const dispatch = useDispatch()

    const cart = useSelector(st => st.cart)
    const wishlist = useSelector(st => st.wishlist)


    function checkInCart(id) {
        if(cart.length ===0) return false
        return cart.some(item => item.id === id);

    }
    function checkInWishlist(id) {
        if (wishlist.length === 0) return false
        return wishlist.some(item => item.id === id);

    }


    const { _token, username } = useContext(userContext).current_user
    const key = uuid()

    async function cartEvent(e) {
        const product_id = e.target.parentElement.id || e.target.id
        const action = e.target.parentElement.name || e.target.name
        const type = action === "add" ? "ADD_TO_CART" : "REMOVE_FORM_CART"
        const message = action === "add" ? "Added To Cart" : "Removed From Cart"
        const backgroundColor = action === "add" ? "#FFB500" : "#F93800"

        const data = { _token, username, product_id, action }

        try {
            await ServerApi.cartAction( data)
            flashControl(message, backgroundColor)

                dispatch({type,product })
        } catch (e) {
            console.log(e)
        }

    }

    async function wishlistEvent(e) {
        const action = e.target.parentElement.name || e.target.name
        const type = action === "add" ? "ADD_TO_WISHLIST" : "REMOVE_FORM_WISHLIST"
        const message = action === "add" ? "Added To Wishlist" : "Removed From Wishlist"
        const backgroundColor = action === "add" ? "#FFB500" : "#F93800"
        const product_id = product.id
        console.log(e.target.id, product.id, action, type, username, _token)
        const data = { _token, username, product_id, action }
        // call server
        try {
            await ServerApi.wishlistAction(data)
            flashControl(message,backgroundColor)
                dispatch({type,product })

        } catch (e) {
            console.log(e)
        }

    }

    function flashControl(message, backgroundColor) {
        setFlash(data =>data = {
            condition: true,
            message ,
            backgroundColor

            })
            setTimeout(() => {
                setFlash(data =>data = {
                    condition: false,
                    message: "",
                    backgroundColor: ""

                    })
            }, 3000)
    }



    return (
        <div className="card-wrapper"  key={key}>
                <div className="product-img-wrapper">
                    <img className="product-img" src={product.image} alt="product-display"></img>
                </div>

                <div className="product-info" >
                    <div className="product-title" >
                        <h5>{product.title}</h5>
                    </div>
                    <div className="product-descripition">
                        <p>{product.description}</p>
                    </div>
                    <div className='product-price-buy' >
                    <b >{`Price: ${product.price}`}$</b>

                    <button id={product.id} name="add" onClick={cartEvent} >
                        <i name="add" className="fas fa-cart-plus" ></i>
                    </button>
                    {checkInCart(product.id) &&
                        <button id={product.id} name="remove" onClick={cartEvent} >
                             <i  name="remove" className="fas fa-minus-square" ></i>
                        </button>
                    }
                    {checkInWishlist(product.id) ?

                        <button name= "remove" onClick={wishlistEvent}>
                        <i name= "remove" className="fas fa-heart"></i>
                        </button> :
                        <button name="add"  onClick={wishlistEvent} >
                        <i  name="add" className="far fa-heart"></i>
                        </button>

                }
                </div>
                {flash.condition && <Alert style={{backgroundColor:flash.backgroundColor}} className="alert">{flash.message}</Alert>}
                </div>

            </div>
    )
}