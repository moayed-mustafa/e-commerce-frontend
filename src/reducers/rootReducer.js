
import { combineReducers } from 'redux'
import productsReducer from './products'
import cartReducer from './cart'
import wishlistReducer from './wishlist'
const rootReducer = combineReducers({products:productsReducer,
    cart: cartReducer,
    wishlist: wishlistReducer
})
export default rootReducer;

