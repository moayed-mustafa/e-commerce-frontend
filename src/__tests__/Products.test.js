
import React from "react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { Route, MemoryRouter } from "react-router-dom";


import rootReducer from '../reducers/rootReducer';
import ProductCard from "../ProductCard";
import UserContext from '../userContext'



let store = createStore(rootReducer);
const product = {
    category: "men clothing",
description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
id: 2,
image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
price: 22.3,
title: "Mens Casual Premium Slim Fit T-Shirts "
}

let current_user= {
    username: "carter", _token:"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImNhcnRlciIsImlhdCI6MTYwNTMwMjY3Nn0.MsDqmb1zxzzHTsI38EwdYJoOIWIMcJGmN1XL0Tm2Trw"

}

function checkUser(user) {
    if (!user || !user._token) {
        localStorage.removeItem("user");

    }else {
            localStorage.setItem("user",JSON.stringify(user))
    }

}

describe('<ProductCard/>', () => {

    it('passes smoke test', function() {
        render((
            <UserContext.Provider value={{ current_user, set_current_user: checkUser }}>
                <MemoryRouter>
                <Provider store={store}>
                    <Route>
                        <ProductCard product={ product}/>
                    </Route>
                </Provider>
                </MemoryRouter>

            </UserContext.Provider>
        ));
    });

    it('passes snapshot test', function() {
        const { asFragment } = render((
            <UserContext.Provider value={{current_user, set_current_user:checkUser}}>
                 <MemoryRouter>
                <Provider store={store}>
                    <Route>
                            <ProductCard product={ product}/>
                    </Route>
                </Provider>
                </MemoryRouter>
            </UserContext.Provider>
        ));
        expect( asFragment() ).toMatchSnapshot();
    });

});
