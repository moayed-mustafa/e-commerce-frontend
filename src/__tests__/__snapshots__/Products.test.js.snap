// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard/> passes snapshot test 1`] = `
<DocumentFragment>
  <div
    class="card-wrapper"
  >
    <div
      class="product-img-wrapper"
    >
      <img
        alt="product-display"
        class="product-img"
        src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
      />
    </div>
    <div
      class="product-info"
    >
      <div
        class="product-title"
      >
        <h5>
          Mens Casual Premium Slim Fit T-Shirts 
        </h5>
      </div>
      <div
        class="product-descripition"
      >
        <p>
          Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.
        </p>
      </div>
      <div
        class="product-price-buy"
      >
        <b>
          Price: 22.3$
        </b>
        <button
          id="2"
          name="add"
        >
          <i
            class="fas fa-cart-plus"
            name="add"
          />
        </button>
        <button
          name="add"
        >
          <i
            class="far fa-heart"
            name="add"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Products/> passes snapshot test 1`] = `
<DocumentFragment>
  <div
    aria-busy="true"
    class=""
  >
    <svg
      aria-label="audio-loading"
      fill="#F93800"
      height="250"
      viewBox="0 0 120 30"
      width="250"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="15"
        cy="15"
        r="15"
      >
        <animate
          attributeName="r"
          begin="0s"
          calcMode="linear"
          dur="0.8s"
          from="15"
          repeatCount="indefinite"
          to="15"
          values="15;9;15"
        />
        <animate
          attributeName="fillOpacity"
          begin="0s"
          calcMode="linear"
          dur="0.8s"
          from="1"
          repeatCount="indefinite"
          to="1"
          values="1;.5;1"
        />
      </circle>
      <circle
        attributeName="fillOpacity"
        cx="60"
        cy="15"
        from="1"
        r="9"
        to="0.3"
      >
        <animate
          attributeName="r"
          begin="0s"
          calcMode="linear"
          dur="0.8s"
          from="9"
          repeatCount="indefinite"
          to="9"
          values="9;15;9"
        />
        <animate
          attributeName="fillOpacity"
          begin="0s"
          calcMode="linear"
          dur="0.8s"
          from="0.5"
          repeatCount="indefinite"
          to="0.5"
          values=".5;1;.5"
        />
      </circle>
      <circle
        cx="105"
        cy="15"
        r="15"
      >
        <animate
          attributeName="r"
          begin="0s"
          calcMode="linear"
          dur="0.8s"
          from="15"
          repeatCount="indefinite"
          to="15"
          values="15;9;15"
        />
        <animate
          attributeName="fillOpacity"
          begin="0s"
          calcMode="linear"
          dur="0.8s"
          from="1"
          repeatCount="indefinite"
          to="1"
          values="1;.5;1"
        />
      </circle>
    </svg>
  </div>
</DocumentFragment>
`;
